File: API/0014_TradingView_Supertrend_Flip/CS/TradingViewSupertrendFlipStrategy.cs
using System;
using System.Collections.Generic;

using Ecng.Common;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// Strategy based on Supertrend indicator flips.
/// Detects when Supertrend direction changes and trades accordingly.
/// </summary>
public class TradingViewSupertrendFlipStrategy : Strategy
{
	private readonly StrategyParam<int> _supertrendPeriod;
	private readonly StrategyParam<decimal> _supertrendMultiplier;
	private readonly StrategyParam<DataType> _candleType;

	private decimal _prevSupertrendValue;
	private bool _prevIsUpTrend;
	private bool _hasPrevValues;

	/// <summary>
	/// Period for Supertrend calculation.
	/// </summary>
	public int SupertrendPeriod
	{
		get => _supertrendPeriod.Value;
		set => _supertrendPeriod.Value = value;
	}

	/// <summary>
	/// Multiplier for Supertrend calculation.
	/// </summary>
	public decimal SupertrendMultiplier
	{
		get => _supertrendMultiplier.Value;
		set => _supertrendMultiplier.Value = value;
	}

	/// <summary>
	/// Candle type.
	/// </summary>
	public DataType CandleType
	{
		get => _candleType.Value;
		set => _candleType.Value = value;
	}

	/// <summary>
	/// Initialize the TradingView Supertrend Flip strategy.
	/// </summary>
	public TradingViewSupertrendFlipStrategy()
	{
		_supertrendPeriod = Param(nameof(SupertrendPeriod), 10)
			.SetDisplay("Supertrend Period", "Period for Supertrend calculation", "Indicators")
			.SetOptimize(7, 14, 1);

		_supertrendMultiplier = Param(nameof(SupertrendMultiplier), 4.0m)
			.SetDisplay("Supertrend Multiplier", "Multiplier for Supertrend", "Indicators")
			.SetOptimize(3.0m, 5.0m, 0.5m);

		_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(5).TimeFrame())
			.SetDisplay("Candle Type", "Type of candles to use", "General");
	}

	/// <inheritdoc />
	public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
	{
		return [(Security, CandleType)];
	}

	/// <inheritdoc />
	protected override void OnReseted()
	{
		base.OnReseted();
		_prevSupertrendValue = default;
		_prevIsUpTrend = default;
		_hasPrevValues = default;
	}

	/// <inheritdoc />
	protected override void OnStarted2(DateTime time)
	{
		base.OnStarted2(time);

		var supertrend = new SuperTrend
		{
			Length = SupertrendPeriod,
			Multiplier = SupertrendMultiplier
		};

		var subscription = SubscribeCandles(CandleType);
		subscription
			.Bind(supertrend, ProcessCandle)
			.Start();

		var area = CreateChartArea();
		if (area != null)
		{
			DrawCandles(area, subscription);
			DrawIndicator(area, supertrend);
			DrawOwnTrades(area);
		}
	}

	private void ProcessCandle(ICandleMessage candle, decimal supertrendValue)
	{
		if (candle.State != CandleStates.Finished)
			return;

		if (!IsFormedAndOnlineAndAllowTrading())
			return;

		if (supertrendValue == 0)
			return;

		// Determine trend direction
		var isUpTrend = candle.ClosePrice > supertrendValue;

		if (!_hasPrevValues)
		{
			_hasPrevValues = true;
			_prevIsUpTrend = isUpTrend;
			_prevSupertrendValue = supertrendValue;
			return;
		}

		// Detect flip
		var isFlippedBullish = isUpTrend && !_prevIsUpTrend;
		var isFlippedBearish = !isUpTrend && _prevIsUpTrend;

		_prevIsUpTrend = isUpTrend;
		_prevSupertrendValue = supertrendValue;

		if (isFlippedBullish && Position <= 0)
		{
			var volume = Volume + Math.Abs(Position);
			BuyMarket(volume);
		}
		else if (isFlippedBearish && Position >= 0)
		{
			var volume = Volume + Math.Abs(Position);
			SellMarket(volume);
		}
	}
}

File: API/0014_TradingView_Supertrend_Flip/PY/tradingview_supertrend_flip_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import SuperTrend
from StockSharp.Algo.Strategies import Strategy

class tradingview_supertrend_flip_strategy(Strategy):
    """
    Strategy based on Supertrend indicator flips.
    Detects when Supertrend direction changes and trades accordingly.
    """

    def __init__(self):
        super(tradingview_supertrend_flip_strategy, self).__init__()
        self._supertrend_period = self.Param("SupertrendPeriod", 10).SetDisplay("Supertrend Period", "Period for Supertrend calculation", "Indicators")
        self._supertrend_multiplier = self.Param("SupertrendMultiplier", 4.0).SetDisplay("Supertrend Multiplier", "Multiplier for Supertrend", "Indicators")
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromMinutes(5))).SetDisplay("Candle Type", "Type of candles to use", "General")

        self._prev_supertrend_value = 0.0
        self._prev_is_up_trend = False
        self._has_prev_values = False

    @property
    def candle_type(self):
        return self._candle_type.Value

    def OnReseted(self):
        super(tradingview_supertrend_flip_strategy, self).OnReseted()
        self._prev_supertrend_value = 0.0
        self._prev_is_up_trend = False
        self._has_prev_values = False

    def OnStarted(self, time):
        super(tradingview_supertrend_flip_strategy, self).OnStarted(time)

        supertrend = SuperTrend()
        supertrend.Length = self._supertrend_period.Value
        supertrend.Multiplier = self._supertrend_multiplier.Value

        subscription = self.SubscribeCandles(self.candle_type)
        subscription.Bind(supertrend, self._process_candle).Start()

        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawIndicator(area, supertrend)
            self.DrawOwnTrades(area)

    def _process_candle(self, candle, st_val):
        if candle.State != CandleStates.Finished:
            return
        if not self.IsFormedAndOnlineAndAllowTrading():
            return

        sv = float(st_val)
        if sv == 0:
            return

        is_up_trend = float(candle.ClosePrice) > sv

        if not self._has_prev_values:
            self._has_prev_values = True
            self._prev_is_up_trend = is_up_trend
            self._prev_supertrend_value = sv
            return

        flipped_bullish = is_up_trend and not self._prev_is_up_trend
        flipped_bearish = not is_up_trend and self._prev_is_up_trend

        self._prev_is_up_trend = is_up_trend
        self._prev_supertrend_value = sv

        if flipped_bullish and self.Position <= 0:
            self.BuyMarket()
        elif flipped_bearish and self.Position >= 0:
            self.SellMarket()

    def CreateClone(self):
        return tradingview_supertrend_flip_strategy()
